traffic_portal/app/src/modules/public/login/LoginController.js

```diff
@@ -48,9 +53,31 @@ var LoginController = function($scope, $log, $uibModal, authService, userService
         });
     };
 
+    $scope.loginOauth = function() {
+        var continueURL = '';
+        const gotoUrl = propertiesModel.properties.oAuth.oAuthUrl;
+        const gotoUrlWithoutParams = gotoUrl.split('?')[0];
+        const queryParams = urlUtils.getUrlQueryParams(window.location.href);
+        const goToQueryParams = urlUtils.getUrlQueryParams(gotoUrl);
+        var returnParams = '?redirect=' + queryParams['redirect'];
+        // keeps query parameters that were added to the goToUrl
+        for (var i in goToQueryParams) {
+            returnParams = returnParams + '&' + i + '=' + goToQueryParams[i];
+        }
+
+        if (returnParams) {
+            continueURL = '?continue=' + $scope.returnProtocol + '//' + $scope.returnUrl + ($scope.returnPort ? ':' + $scope.returnPort : '') + '/' + encodeURIComponent('#!') + '/sso' + returnParams;
+        } else {
+            continueURL = '?continue=' + $scope.returnProtocol + '//' + $scope.returnUrl + ($scope.returnPort ? ':' + $scope.returnPort : '') + '/' + encodeURIComponent('#!') + '/sso';
```

Comment:
instead of separating the parts of `window.location`, why not store the full `Location` object so you could just manipulate the `hash` property thereof? That's essentially what you're doing, but currently storing the information in three different places and doing inline collapsing and ternaries and whatnot (that will also handle URI encoding for you).

Reply:
done